README.md
This is the README file for MTK-MYB-CNC,
MTK-MYB CNC implementation.

## Description

MTK-MYB-CNC provides the MTK-MYB CNC implementation.

This allows a single host to backup any number of mysql servers
on the local network without having to maintain instances of a backup
script on these hosts. It only requires a single host to be configured
and maintained. The hosts being backed up only need curl installed
and they must be configured to allow password less ssh access
from the central backup instance.

The central host uses pure-ftpd to retrieve the backups from the
remote hosts. It connects to them via SSH and issues a mysqldump
command for each of the tables. The output from mysqldump is then
compressed on the fly and directly fed to STDIN of curl. So no
data is written to disk on the hosts being backed up. This is usually
the fastest and most convenient way to take backups from the backup hosts.

The backup hosts will be locked during the time of the backup, so these
should be replication slaves of a mysql master.

## Installation

This package uses Dist::Zilla.

Use

dzil build

to create a release tarball which can be
unpacked and installed like any other EUMM
distribution.

perl Makefile.PL

make

make test

make install

## Documentation

Please see perldoc MTK::MYB::CNC.


Changes
Revision history for MTK-MYB-CNC

0.19    2012-12-26
        First public version, released on an unsuspecting world.

lib/MTK/MYB/CNC/Job.pm
package MTK::MYB::CNC::Job;
# ABSTRACT: MTK CNC MYB job

use 5.010_000;
use mro 'c3';
use feature ':5.10';

use Moose;
use namespace::autoclean;

# use IO::Handle;
# use autodie;
# use MooseX::Params::Validate;
# use Carp;
# use English qw( -no_match_vars );
# use Try::Tiny;

use MTK::MYB::CNC::Worker;

extends 'MTK::MYB::Job';

# has ...
# with ...
# initializers ...

sub _init_worker {
    my $self = shift;

    my $Worker = MTK::MYB::CNC::Worker::->new(
        {
            'config'  => $self->config(),
            'logger'  => $self->logger(),
            'parent'  => $self->parent(),
            'verbose' => $self->verbose(),
            'dry'     => $self->dry(),
            'bank'    => $self->bank(),
            'vault'   => $self->vault(),
        }
    );

    return $Worker;
}

no Moose;
__PACKAGE__->meta->make_immutable;

1;

__END__

=head1 NAME

MTK::MYB::CNC::Job - MTK CNC MYB job

=head1 ACKNOWLEDGEMENT

This module was originally developed for eGENTIC Systems. With approval from eGENTIC Systems,
this module was generalized and published, for which the authors would like to express their
gratitude.

=cut


lib/MTK/MYB/CNC/Worker.pm
package MTK::MYB::CNC::Worker;
# ABSTRACT: CNC MYB worker

use 5.010_000;
use mro 'c3';
use feature ':5.10';

use Moose;
use namespace::autoclean;

# use IO::Handle;
# use autodie;
# use MooseX::Params::Validate;

use MTK::MYB::Codes;

extends 'MTK::MYB::Worker';

has 'curl_exitcodes' => (
    'is'      => 'ro',
    'isa'     => 'HashRef',
    'lazy'    => 1,
    'builder' => '_init_curl_exitcodes',
);

sub _init_curl_exitcodes {
    my $self = shift;

    my %codes = (
        0 => 'OK',
        1 => 'Unsupported protocol. This build of curl has no support for this protocol.',
        2 => 'Failed to initialize.',
        3 => 'URL malformed. The syntax was not correct.',
        5 => "Couldn't resolve proxy.  The  given  proxy  host  could  not  be resolved.",
        6 => "Couldn't resolve host. The given remote host was not resolved.",
        7 => 'Failed to connect to host.',
        8 => "FTP  weird  server  reply.  The  server  sent data curl couldn't parse.",
        9 =>
"FTP access denied. The server denied login or denied  access  to the  particular  resource or directory you wanted to reach. Most often you tried to change to a directory that doesn't  exist  on the server.",
        11 => "FTP  weird PASS reply. Curl couldn't parse the reply sent to the PASS request.",
        13 => "FTP weird PASV reply, Curl couldn't parse the reply sent to  the PASV request.",
        14 => "FTP  weird  227  format.  Curl  couldn't  parse the 227-line the server sent.",
        15 => "FTP can't get host. Couldn't resolve the host IP we got  in  the 227-line.",
        17 => "FTP  couldn't  set  binary.  Couldn't  change transfer method to binary.",
        18 => "Partial file. Only a part of the file was transferred.",
        19 => "FTP couldn't download/access the given file, the RETR (or  simi‐lar) command failed.",
        21 => "FTP quote error. A quote command returned error from the server.",
        22 =>
"HTTP  page  not  retrieved.  The  requested url was not found or returned another error with the HTTP error  code  being  400  or above. This return code only appears if -f/--fail is used.",
        23 => "Write  error.  Curl couldn't write data to a local filesystem or similar.",
        25 => "FTP couldn't STOR file. The server denied  the  STOR  operation, used for FTP uploading.",
        26 => "Read error. Various reading problems.",
        27 => "Out of memory. A memory allocation request failed.",
        28 => "Operation  timeout.  The  specified  time-out period was reached according to the conditions.",
        30 => "FTP PORT failed. The PORT command failed. Not  all  FTP  servers support  the  PORT  command,  try  doing  a  transfer using PASV instead!",
        31 => "FTP couldn't use REST. The REST command failed. This command  is used for resumed FTP transfers.",
        33 => "HTTP range error. The range 'command' didn't work.",
        34 => "HTTP post error. Internal post-request generation error.",
        35 => "SSL connect error. The SSL handshaking failed.",
        36 => "FTP  bad  download  resume. Couldn't continue an earlier aborted download.",
        37 => "FILE couldn't read file. Failed to open the file. Permissions?",
        38 => "LDAP cannot bind. LDAP bind operation failed.",
        39 => "LDAP search failed.",
        41 => "Function not found. A required LDAP function was not found.",
        42 => "Aborted by callback. An application told curl to abort the oper‐ation.",
        43 => "Internal error. A function was called with a bad parameter.",
        45 => "Interface  error.  A  specified  outgoing interface could not be used.",
        47 => "Too many redirects. When following redirects, curl hit the maxi‐mum amount.",
        48 => "Unknown TELNET option specified.",
        49 => "Malformed telnet option.",
        51 => "The peer's SSL certificate or SSH MD5 fingerprint was not ok.",
        52 => "The  server  didn't  reply anything, which here is considered an error.",
        53 => "SSL crypto engine not found.",
        54 => "Cannot set SSL crypto engine as default.",
        55 => "Failed sending network data.",
        56 => "Failure in receiving network data.",
        58 => "Problem with the local certificate.",
        59 => "Couldn't use specified SSL cipher.",
        60 => "Peer certificate cannot be authenticated with known CA  certifi‐cates.",
        61 => "Unrecognized transfer encoding.",
        62 => "Invalid LDAP URL.",
        63 => "Maximum file size exceeded.",
        64 => "Requested FTP SSL level failed.",
        65 => "Sending the data requires a rewind that failed.",
        66 => "Failed to initialise SSL Engine.",
        67 => "The  user  name, password, or similar was not accepted and curl failed to log in.",
        68 => "File not found on TFTP server.",
        69 => "Permission problem on TFTP server.",
        70 => "Out of disk space on TFTP server.",
        71 => "Illegal TFTP operation.",
        72 => "Unknown TFTP transfer ID.",
        73 => "File already exists (TFTP).",
        74 => "No such user (TFTP).",
        75 => "Character conversion failed.",
        76 => "Character conversion functions required.",
        77 => "Problem with reading the SSL CA cert (path? access rights?).",
        78 => "The resource referenced in the URL does not exist.",
        79 => "An unspecified error occurred during the SSH session.",
        80 => "Failed to shut down the SSL connection.",
        82 => "Could not load CRL file,  missing  or  wrong  format.",
        83 => "Issuer check failed",
    );

    return \%codes;
}

sub uid {
    my $self = shift;

    return $self->parent()->ftp_uid();
}

sub gid {
    my $self = shift;

    return $self->parent()->ftp_gid();
}

sub type {
    return 'cnc';
}

sub _prepare {
    my $self = shift;
    my $host = shift;
    my $dbms = shift;

    # turn of ssh host key checking
    $self->sys()->ssh_hostkey_check(0);

    my $server_address = $self->config()->get_scalar('MTK::MYB::CNC::ServerAddress');

    if ( !$server_address ) {
        $self->logger()->log(
            message =>
              'No server address given. You need to set THIS hosts external (wrt. your backup nodes) address to the key MTK::MYB::CNC::ServerAddress',
            level => 'error',
        );
        return;
    }

    if ( !$host ) {
        $self->logger()->log( message => "No host given! Impossible to continue.", level => 'error', );
        return;
    }

    if ( !$dbms ) {
        $self->logger()->log( message => "No dbms given! Impossible to continue.", level => 'error', );
        return;
    }

    # check if pw-less ssh access works
    if ( !$self->sys()->run_remote_cmd( $host, '/bin/true' ) ) {

        # report an error is pw-less ssh access does not work
        my $msg = 'Password-less SSH access to ' . $host . ' does not work. Public Keys setup ok? Aborting!';
        $self->logger()->log( message => $msg, level => 'error', );
        $self->status()->global( MTK::MYB::Codes::get_status_code('SSH-ERROR') );
        $self->_report();
        return;
    }
    else {
        $self->logger()->log( message => 'Password-less SSH access to '.$host.' is OK', level => 'debug', );
    }

    # check ftp connection via ssh
    # lftp -u $dbms,$config->{$dbms}{'ftp-password'} $config->{'default'}{'server-address'} -e "quit"
    my $rv = $self->sys()->run_remote_cmd( $host, $self->_get_ftp_cmd_test($dbms), { ReturnRV => 1, } );
    if ( defined($rv) && $rv == 0 ) {
        $self->logger()->log( message => 'FTP access from ' . $host . ' is OK', level => 'debug', );
    }
    else {

        # report an error if ftp-access from client does not work
        my $addn_error_msg = '';
        if ( $self->parent()->ftp_helper() eq 'curl' ) {
            my $error_msg = $self->curl_exitcodes()->{$rv} || 'n/a';
            $addn_error_msg = 'Curl exited with Code ' . $rv . ' which means: "' . $error_msg . '". ';
        }
        my $msg = 'FTP access from ' . $host . ' to ' . $server_address . ' does not work. ' . $addn_error_msg . ' Is lftp/curl installed? Aborting!';
        $self->logger()->log( message => $msg, level => 'error', );
        $self->status()->global( MTK::MYB::Codes::get_status_code('FTP-ERROR') );
        $self->_report();
        return;
    }

    return 1;
}

sub _binlog_archive {
    my $self = shift;

    my $host = $self->hostname();
    $self->logger()->log( message => 'binlog_archive processing '.$host, level => 'debug', );

    # Create final archive destination
    my $local_archive_dir = $self->dir_binlogs();

    # Create temporary transfer destination
    my $local_copy_dir = $self->fs()->makedir( $self->fs()->filename( $self->dir_progress(), 'binlogs' ), { Uid => $self->uid(), Gid => $self->gid(), } );

    # The binlogs must first be copied to the inprogress dir - since only this
    # directory is being exported by the ftpd - and the copied
    # to their final destination.

    # get remote log-bin setting
    my $cmd = 'egrep "log(_|-)bin" /etc/mysql/my.cnf | cut -d"=" -f2';
    my $remote_log_bin = $self->sys()->run_remote_cmd( $host, $cmd, { CaptureOutput => 1, Chomp => 1, } );
    my $remote_source_dir = File::Basename::dirname($remote_log_bin);

    # default = 10 years
    my $holdbacktime_binlog = $self->config()->get('MTK::MYB::Rotations::Binlogs') || 3650;

    if ($remote_source_dir) {
        $self->logger()->log( message => 'Using Remote Binlog Source Directory :' . $remote_source_dir, level => 'debug', );
    }
    else {
        $self->logger()->log( message => 'Remote Binlog Source Directory not set! Returning.', level => 'error', );
        return;
    }

    # make sure the temporary copy directory is defined and exists
    if ( $local_copy_dir && -d $local_copy_dir && $local_copy_dir ne '/' ) {
        $self->logger()->log( message => 'Using local copy directory: ' . $local_copy_dir, level => 'debug', );
    }
    else {
        $local_copy_dir ||= '';    # prevents an undefinedness warning
        $self->logger()->log( message => 'Local copy directory not defined or not accessible: ' . $local_copy_dir, level => 'error', );
        return;
    }

    # make sure the final archive destination exists
    if ( $local_archive_dir && -d $local_archive_dir ) {
        $self->logger()->log( message => 'Using local archive directory: ' . $local_archive_dir, level => 'debug', );
    }
    else {
        $local_archive_dir ||= '';    # prevents an undefinedness warning
        $self->logger()->log( message => 'Local archive driectory not defined or not accessible: ' . $local_archive_dir, level => 'error', );
        return;
    }

    my ( $sec, $min, $hour, $dayofmonth, $month, $year, $dayofweek, $dayofyear, $summertime ) = localtime(time);
    $year += 1900;
    $month++;

    # Remove old (expired) archived binlogs [local]
    foreach my $file ( glob( $local_archive_dir . '/mysql-bin.*' ) ) {
        $self->logger()->log( message => 'Binlogarchive-Expire - File '.$file.' is ' . sprintf( '%.2f', -M $file ) . ' days old.', level => 'debug', );
        if ( -M $file > $holdbacktime_binlog ) {
            $self->logger()->log(
                message => 'Binlogarchive-Expire - File ' . $file . ' is too ' . sprintf( '%.2f', -M $file ) . ' old. Removing.',
                level   => 'debug',
            );
            $cmd = 'rm -f '.$file;
            $self->logger()->log( message => $cmd, level => 'debug', );
            $self->sys()->run_cmd($cmd) unless $self->dry();
        }
    }

    # Archive new binlogs [remote]
    if ( $remote_source_dir && $remote_source_dir ne q{/} ) {
        $self->logger()->log( message => 'Continuing w/ remote binlog search path '.$remote_source_dir, level => 'debug', );
    }
    else {
        $self->logger()->log( message => 'Invalid remote binlog search path: '.$remote_source_dir, level => 'error', );
        return;
    }

    $cmd = 'find ' . $remote_source_dir . ' -type f -name "mysql-bin.*"';
    my $out = $self->sys()->run_remote_cmd( $host, $cmd, { CaptureOutput => 1, Timeout => 1200, Chomp => 1, } );

    $out ||= '';    # prevent definedness warnings
    foreach my $binlog_file ( split /\n/, $out ) {
        $self->logger()->log( message => 'Examining Binlog File: ' . $binlog_file, level => 'debug', );

        my @srcpath = split /\//, $binlog_file;
        my $dst = $srcpath[-1] . $self->packer()->ext();
        if ( !-e $dst || -M $dst < 1 ) {
            if ( !-e $dst ) {
                $self->logger()->log( message => $binlog_file.' not present, creating '.$dst, level => 'debug', );
            }
            else {
                $self->logger()->log( message => $binlog_file.' not up-to-date, overwriting '.$dst, level => 'debug', );
            }
            $cmd = $self->_get_cmd_prefix();
            $cmd .= $self->packer()->cmd();
            $cmd .= q{ } . $binlog_file;
            $cmd .= q{ | };
            $cmd .= $self->_get_cmd_prefix() . $self->_get_ftp_cmd_stdin( 'binlogs', $dst );
            $self->logger()->log( message => 'CMD: ' . $cmd, level => 'debug', );
            $self->sys()->run_remote_cmd( $host, $cmd, {} );
        }
        else {
            $self->logger()->log( message => 'Binlogfile at ' . $dst . ' present and uptodate, skipping.', level => 'debug', );
        }
    }

    # move archived binlogs from the incoming copy directory to their permanent storage location
    $cmd = q{mv -f }.$local_copy_dir.q{/* }.$local_archive_dir.q{/};
    if ( $self->sys()->run_cmd( $cmd, { Timeout => 3600, } ) ) {
        $self->logger()
          ->log( message => 'Moved binlogs from transfer directory at '.$local_copy_dir.' to final archive destination at '.$local_archive_dir, level => 'debug', );
        $cmd = q{rm -rf }.$local_copy_dir.q{/};
        if ( $self->sys()->run_cmd( $cmd, { Timeout => 3600, } ) ) {
            $self->logger()->log( message => 'Removed temporary transfer directory ' . $local_copy_dir, level => 'debug', );
        }
        else {
            $self->logger()->log( message => 'Failed to remove temporary transfer directory ' . $local_copy_dir, level => 'error', );
        }
    }
    else {
        $self->logger()->log(
            message => 'Failed to move binlogs from transfer directory at '.$local_copy_dir.' to final archive destination at '.$local_archive_dir,
            level   => 'error',
        );
    }

    # remove old binlogs [remote]
    $cmd = $self->_get_cmd_prefix();
    $cmd .= '/usr/bin/find ' . $remote_source_dir . ' -type f -regex ".*mysql-bin.[0-9].*" -mtime +10 -print0 | /usr/bin/xargs -0 rm -f';
    $self->logger()->log( message => $cmd, level => 'debug', );
    if ( $self->sys()->run_remote_cmd( $host, $cmd, ) ) {
        $self->logger()->log( message => 'Removed old binlogs on remote host '.$host, level => 'debug', );
        return 1;
    }
    else {
        $self->logger()->log( message => 'Failed to remove old binlogs on remote host '.$host, level => 'debug', );
        return;
    }
}

sub _get_cmd_suffix {
    my $self    = shift;
    my $db      = shift;
    my $table   = shift;
    my $type    = shift;
    my $destdir = shift;
    my $file    = shift;

    my $cmd = ' | ' . $self->_get_cmd_prefix() . $self->_get_ftp_cmd_base('--ftp-create-dirs -T -');
    if ( $self->parent()->ftp_helper() eq 'curl' ) {
        $cmd .= q{/} . $type . q{/} . $db . q{/} . $file . $self->packer()->ext();
    }
    else {
        $cmd .= ' -e "cd ' . $type . '; cd ' . $db . '; put /dev/stdin -o ' . $file . $self->packer()->ext() . '; quit"';
    }

    return $cmd;
}

sub _get_ftp_cmd_base {
    my $self = shift;
    my $opts = shift || '';

    my $dbms = $self->dbms();

    my $password       = $self->config()->get( 'MTK::MYB::DBMS::' . $dbms . '::Password' );
    my $server_address = $self->config()->get('MTK::MYB::CNC::ServerAddress');
    my $ftp_cmd_base;

    if ( $self->parent()->ftp_helper() eq 'curl' ) {
        $ftp_cmd_base = '/usr/bin/curl ' . $opts . ' -s -4 -u ' . $dbms . q{:} . $password . ' ftp://' . $server_address . q{:} . $self->parent()->ftp_port();
    }
    else {
        $ftp_cmd_base = '/usr/bin/lftp ' . $opts . q{ -p } . $self->parent()->ftp_port() . q{ -u } . $dbms . q{,"} . $password . q{" } . $server_address;
    }

    return $ftp_cmd_base;
}

sub _get_ftp_cmd_stdin {
    my $self      = shift;
    my $dest_dir  = shift;
    my $dest_file = shift;

    if ( $self->parent()->ftp_helper() eq 'curl' ) {
        return $self->_get_ftp_cmd_base('--ftp-create-dirs -T -') . q{/} . $dest_dir . q{/} . $dest_file;
    }
    else {
        return $self->_get_ftp_cmd_base() . ' -e "cd ' . $dest_dir . '; put /dev/stdin -o ' . $dest_file . '; quit"';
    }
}

sub _get_ftp_cmd_test {
    my $self = shift;

    if ( $self->parent()->ftp_helper() eq 'curl' ) {
        return $self->_get_ftp_cmd_base() . q{/};
    }
    else {
        return $self->_get_ftp_cmd_base() . '-e "quit"';
    }
}

no Moose;
__PACKAGE__->meta->make_immutable;

1;

__END__

=head1 NAME

MTK::MYB::CNC::Worker - The Workhorse for the centralized (CNC) MySQLBackup

=head1 SYNOPSIS

    use MTK::MYB::CNC::Worker;
    my $Worker = MTK::MYB::CNC::Worker::->new(
        {
            'config'  => $self->config(),
            'logger'  => $self->logger(),
            'parent'  => $self->parent(),
            'verbose' => $self->verbose(),
            'dry'     => $self->dry(),
            'bank'    => $self->bank(),
            'vault'   => $self->vault(),
        }
    );
    $Worker->run();
    
=head1 DESCRIPTION

This class implements the business logic of the CNC MySQL Backup solution. It
extends MTK::MYB::Worker and overrides some of its methods.

=method run

Once the class has been set up call this method to start the backup process.

=method type

This method is primarily usefull for subclassing this class. It is used to
dynamically determine the exact subtype of itself. Of course this
could as well be done by using ISA and/or ref. Howevery this way is more
straight forward and easier to implement.

=method uid

The ftp uid.

=method gid

The ftp gid.

=head1 ACKNOWLEDGEMENT

This module was originally developed for eGENTIC Systems. With approval from eGENTIC Systems,
this module was generalized and published, for which the authors would like to express their
gratitude.

=cut


lib/MTK/MYB/CNC.pm
package MTK::MYB::CNC;
# ABSTRACT: MTK-MYB CNC implementation

use 5.010_000;
use mro 'c3';
use feature ':5.10';

use Moose;
use namespace::autoclean;

# use IO::Handle;
# use autodie;
# use MooseX::Params::Validate;

use Crypt::PasswdMD5 qw();
use Data::Dumper;
use Data::Pwgen;
use MTK::MYB::CNC::Job;

extends 'MTK::MYB';

has 'ftp_helper' => (
    'is'      => 'ro',
    'isa'     => 'Str',
    'default' => 'curl',
);

has 'ftp_port' => (
    'is'      => 'ro',
    'isa'     => 'Int',
    'lazy'    => 1,
    'builder' => '_init_ftp_port',
);

has 'ftp_uid' => (
    'is'      => 'ro',
    'isa'     => 'Int',
    'lazy'    => 1,
    'builder' => '_init_ftp_uid',
);

has 'ftp_gid' => (
    'is'      => 'ro',
    'isa'     => 'Int',
    'lazy'    => 1,
    'builder' => '_init_ftp_gid',
);

has 'tempdir' => (
    'is'      => 'ro',
    'isa'     => 'File::Temp::Dir',
    'lazy'    => 1,
    'builder' => '_init_tempdir',
);

sub _plugin_base_class { return ('MTK::MYB::Plugin','MTK::MYB::CNC::Plugin') }

sub _init_tempdir {
    my $self = shift;

    my $Dir = File::Temp::->newdir();

    if ( $self->debug() ) {
        $Dir->unlink_on_destroy(0);
    }

    return $Dir;
}

sub _init_ftp_uid {
    my $self = shift;

    # Try different UIDs until we find one that exists on this host:
    # - ftp-uid should be configured in the config
    # - 34 is the user 'backup'
    # - 1 is the user 'daemon',
    # - 0 is the user 'root',
    my @uids = ( $self->config()->get('MTK::MYB::CNC::ftp-uid'), 34, 1, 0 );
    foreach my $uid (@uids) {
        if ( getpwuid($uid) ) {
            return $uid;
        }
    }

    return 0;
}

sub _init_ftp_gid {
    my $self = shift;

    # Try differents GIDs until we find one that exists on this host:
    # - ftp-gid should be configured in the config
    # - 34 is the group 'backup',
    # - 1 is the group 'daemon',
    # - 0 is the group 'root',
    my @gids = ( $self->config()->get('MTK::MYB::CNC::ftp-gid'), 34, 1, 0 );
    foreach my $gid (@gids) {
        if ( getgrgid($gid) ) {
            return $gid;
        }
    }

    return 0;
}

sub type {
    return 'cnc';
}

sub _prepare {
    my $self = shift;

    $self->SUPER::_prepare()
      or return;
    $self->_stop_ftpd();
    $self->_start_ftpd();
    return 1;
}

sub _cleanup {
    my $self = shift;
    my $ok   = shift;

    if ( $self->debug() ) {
        $self->logger()->log( message => "DEBUG-MODE enabled. Sleeping for 180s before performing cleanup", level => 'debug', );
        sleep 180;
    }

    $self->SUPER::_cleanup($ok);
    $self->_stop_ftpd();

    return 1;
}

sub _get_backup_host {
    my $self = shift;
    my $dbms = shift;

    # host MUST NOT be taken from the default!
    return $self->config()->get( 'MTK::MYB::DBMS::' . $dbms . '::Hostname' );
}

sub _get_job {
    my $self  = shift;
    my $vault = shift;

    my $Job = MTK::MYB::CNC::Job::->new(
        {
            'parent' => $self,
            'vault'  => $vault,
            'logger' => $self->logger(),
            'config' => $self->config(),
            'bank'   => $self->bank(),
        }
    );

    return $Job;
}

sub _init_ftp_port {
    my $self = shift;

    my $ftp_port = $self->config()->get('MTK::MYB::CNC::FtpPort') || 2334;

    return $ftp_port;
}

=head2 PURE-FTPD OPTIONS

This section describes the pure-ftpd options use below.

=over 4

=item -d -d

Turns on debug logging including responses.

=item -S ,<PORT>

Bind to the specified port. The colon is important!

=item -f none

Disable syslog logging. We use our own logfile for that.

=item -j

Auto-create any missing home directories.

=item -z

Allow access to hidden files.

=item -B

Daemonize

=item -H

Don't resolve host names.

=item -4

Use IPv4 only.

=item -c 250

Limit to 250 simultaneous clients.

=item -k 98

Disable uploads if the partition is more than 98% full.

=item -Y 0

Disable all TLS features.

=item -O clf:<LOGFILE>

Enable CLF-style logging to <LOGFILE>.

=item -l puredb:<PDBFILE>

Enable native PureDB authentication from <PDBFILE>.

=back

=cut

sub _start_ftpd {
    my $self = shift;

    my $cleanup = 1;
    if ( $self->debug() ) {
        $cleanup = 0;
    }

    my $tempdir = $self->tempdir()->dirname();

    my $passwdfile = $tempdir . '/pureftpd.passwd';
    if ( $self->_create_pdb($passwdfile) ) {
        $self->logger()->log( message => "Created password-db at $passwdfile", level => 'debug', );
    }
    else {
        $self->logger()->log( message => "Could not create password-db at $passwdfile", level => 'warning', );
        return;
    }

    my $ftp_port = $self->ftp_port();

    my $cmd = "/usr/sbin/pure-ftpd";
    if ( $self->debug() ) {
        $cmd .= " -d -d";
    }
    $cmd .= " -S ,$ftp_port -f none -j -z -B -H -4 -c 250 -k 98 -Y 0 -O clf:$tempdir/ftpd-transfer.log -l puredb:$tempdir/pureftpd.pdb";

    if ( $self->sys()->run_cmd($cmd) ) {
        $self->logger()->log( message => "pure-ftpd started w/ cmd: $cmd", level => 'debug', );
        sleep 15 if $self->config()->get('MTK::MYB::Debug');
        return 1;
    }
    else {
        $self->logger()->log( message => "pure-ftpd failed to start w/ cmd: $cmd", level => 'warning', );
        sleep 15 if $self->config()->get('MTK::MYB::Debug');
        return;
    }
}

#
sub _stash_xfer_log {
    my $self = shift;

    my $tempdir = $self->tempdir()->dirname();

    my $logfile = $tempdir . '/ftpd-transfer.log';
    if ( -e $logfile ) {
        my $target = '/var/log/mysqlbackup-cnc-xfer.log';
        my $cmd    = "cat $logfile >> $target";
        if ( $self->sys()->run_cmd($cmd) ) {
            $self->logger()->log( message => 'Saved pure-ftpd transfer log to ' . $target, level => 'debug', );
        }
        else {
            $self->logger()->log( message => 'Failed to save pure-ftpd transfer log to ' . $target, level => 'error', );
        }
    }
    else {
        $self->logger()->log( message => 'No pure-ftpd transfer log found at ' . $logfile, level => 'notice', );
    }

    my $pdb = $tempdir . '/pureftpd.pdb';
    if ( $self->debug() && -e $pdb ) {
        my $target = '/var/log/mysqlbackup-cnc-pdb-' . time() . '.log';
        my $cmd    = "mv $pdb $target";
        if ( $self->sys()->run_cmd($cmd) ) {
            $self->logger()->log( message => 'Saved pure-ftpd pdb to ' . $target, level => 'debug', );
        }
        else {
            $self->logger()->log( message => 'Failed to save pure-ftpd pdb to ' . $target, level => 'debug', );
        }
    }
    else {
        $self->logger()->log( message => 'Not in debug mode or no pure-ftpd pdb found at ' . $pdb, level => 'debug', );
    }

    return 1;
}

sub _stop_ftpd {
    my $self = shift;

    my ( $cmd, $pid );

    my $ftp_port = $self->ftp_port();

    $cmd = "netstat -nlp | grep pure-ftp | grep \":$ftp_port\" | grep -v grep | awk '{ print \$7; }' | cut -d'/' -f1";
    $self->logger()->log( message => "CMD: $cmd", level => 'debug', );
    $pid = $self->sys()->run_cmd( $cmd, { Chomp => 1, CaptureOutput => 1, } );
    if ( $pid && $pid =~ m/^\s*\d+\s*$/ ) {
        $cmd = "kill $pid >/dev/null 2>&1";
        $self->logger()->log( message => "CMD: $cmd", level => 'debug', );
        $self->sys()->run_cmd($cmd);
        $cmd = "kill -9 $pid >/dev/null 2>&1";
        $self->logger()->log( message => "CMD: $cmd", level => 'debug', );
        $self->sys()->run_cmd($cmd);
    }
    else {
        $self->logger()->log( message => "stop_ftpd - no valid pid found: $pid", level => 'notice', );
    }

    $cmd = "ps aux | grep pure-ftp | grep -v grep | awk '{ print \$2; }'";
    $self->logger()->log( message => "CMD: $cmd", level => 'debug', );
    $self->sys()->run_cmd( $cmd, { Chomp => 1, CaptureOutput => 1, } );

    $self->_stash_xfer_log();

    if ( $pid && $pid =~ m/^\s*\d+\s*$/ ) {
        $cmd = "kill $pid >/dev/null 2>&1";
        $self->logger()->log( message => "CMD: $cmd", level => 'debug', );
        $self->sys()->run_cmd($cmd);
        $cmd = "kill -9 $pid >/dev/null 2>&1";
        $self->logger()->log( message => "CMD: $cmd", level => 'debug', );
        $self->sys()->run_cmd($cmd);
        return 1;
    }
    else {
        $self->logger()->log( message => "stop_ftpd - no valid pid found: $pid", level => 'notice', );
        return;
    }
}

# create pure-db
sub _create_pdb {
    my $self       = shift;
    my $passwdfile = shift;

    my $dbms_ref = $self->config()->get('MTK::MYB::DBMS');

    if(!$dbms_ref || ref($dbms_ref) ne 'HASH') {
        $self->logger()->log( message => 'No DBMS found!', level => 'error', );
        return;
    }

    # DGR: Well, this is brief enough ;)
    ## no critic (RequireBriefOpen)
    if ( open( my $FH, ">", $passwdfile ) ) {
        my $ftp_uid = $self->ftp_uid();
        my $ftp_gid = $self->ftp_gid();

        # create a individual login for each instance
        foreach my $instance ( sort keys %{ $dbms_ref } ) {
            my $salt     = &Data::Pwgen::pwgen( 6,  'alphanum' );
            my $password = &Data::Pwgen::pwgen( 12, 'alphanum' );
            $self->config()->set( 'MTK::MYB::DBMS::' . $instance . '::Password', $password );
            my $crypt = Crypt::PasswdMD5::unix_md5_crypt( $password, $salt );
            my $datadir =
              $self->fs()->makedir( $self->fs()->filename( ( $self->bank(), $instance, 'daily', 'inprogress' ) ), { Uid => $ftp_uid, Gid => $ftp_gid } );

            #my $entry = $instance . ':' . $crypt . ':' . $ftp_uid . ':' . $ftp_gid . '::' . $datadir . '/' . $instance . '::::::::::::' . "\n";
            my $entry = $instance . ':' . $crypt . ':' . $ftp_uid . ':' . $ftp_gid . '::' . $datadir . '::::::::::::' . "\n";
            if ( print $FH $entry ) {
                $self->logger()->log( message => "Passwd-Entry: $entry - l/p: $instance/$password", level => 'debug', );
            }
            else {
                $self->logger()->log( message => "Could not write Passwd-Entry $entry to passwd-file $passwdfile", level => 'warning', );
            }
        }

        if ( !close($FH) ) {
            $self->logger()->log( message => "Could not close file $passwdfile", level => 'debug', );
        }
        ## use critic

        my $pdbfile = $passwdfile;
        $pdbfile =~ s/\.passwd$/.pdb/;
        my $cmd = "/usr/bin/pure-pw mkdb $pdbfile -f $passwdfile";

        if ( $self->sys()->run_cmd($cmd) ) {
            $self->logger()->log( message => "ok mkdb w/ CMD: $cmd", level => 'debug', );
            return 1;
        }
        else {
            $self->logger()->log( message => "failed mkdb w/ CMD: $cmd", level => 3 );
            return;
        }
    }
    else {
        $self->logger()->log( message => "Could not open passwdfile at $passwdfile for writing: $!", level => 'alert' );
        return;
    }
}

no Moose;
__PACKAGE__->meta->make_immutable;

1;

__END__

=head1 NAME

MTK::MYB::CNC - CNC implementation of mysqlbackup

=head1 RISKS

This section is included to help users of this tool to assess the risk associated
with using this tool. The two main categories adressed are those created the idea
implemented and those created by bugs. There may be other risks as well.

B<myb cnc> is mostly a read-only tool that will, however, lock your database server
for the duration of the backup. This will cause service interruptions as long as
you don't take precautions. Either point the script to an dedicated slave or
chose an idle time for running it.

=head1 SEE ALSO

L<MySQL Backup> may be better suited if you plan to back up only a small number
of hosts.

L<Percona XtraBackup|http://www.percona.com/software/percona-xtrabackup> is an
advanced approach for backing up InnoDB and XtraDB tables. It does provide little
advantage in terms of MyISAM backups.

L<Holland Backup|http://wiki.hollandbackup.org/> is an multi-db application
written in Python.

=method type

Always returns cnc.

=head1 ACKNOWLEDGEMENT

This module was originally developed for eGENTIC Systems. With approval from eGENTIC Systems,
this module was generalized and published, for which the authors would like to express their
gratitude.

=cut
